x/stake/handler.go

```diff
@@ -53,212 +61,202 @@ func EndBlocker(ctx sdk.Context, k Keeper) (ValidatorUpdates []abci.Validator) {
-
-	// check if bond has any shares in it unbond
-	bond, found := k.GetDelegation(ctx, msg.DelegatorAddr, msg.ValidatorAddr)
-	if !found {
-		return ErrNoDelegatorForAddress(k.codespace).Result()
+	ubd := UnbondingDelegation{
+		DelegatorAddr: msg.DelegatorAddr,
+		ValidatorAddr: msg.ValidatorAddr,
+		MinTime:       minTime,
+		Balance:       sdk.Coin{params.BondDenom, sdk.NewInt(returnAmount)},
 	}
+	k.SetUnbondingDelegation(ctx, ubd)
 
-	var delShares sdk.Rat
+	tags := sdk.NewTags(
+		tags.Action, tags.ActionBeginUnbonding,
+		tags.Delegator, []byte(msg.DelegatorAddr.String()),
+		tags.SrcValidator, []byte(msg.ValidatorAddr.String()),
+	)
+	return sdk.Result{Tags: tags}
+}
 
-	// test that there are enough shares to unbond
-	if msg.Shares == "MAX" {
-		if !bond.Shares.GT(sdk.ZeroRat()) {
-			return ErrNotEnoughBondShares(k.codespace, msg.Shares).Result()
-		}
-	} else {
-		var err sdk.Error
-		delShares, err = sdk.NewRatFromDecimal(msg.Shares)
-		if err != nil {
-			return err.Result()
-		}
-		if bond.Shares.LT(delShares) {
-			return ErrNotEnoughBondShares(k.codespace, msg.Shares).Result()
-		}
-	}
+func handleMsgCompleteUnbonding(ctx sdk.Context, msg types.MsgCompleteUnbonding, k keeper.Keeper) sdk.Result {
 
-	// get validator
-	validator, found := k.GetValidator(ctx, msg.ValidatorAddr)
+	ubd, found := k.GetUnbondingDelegation(ctx, msg.DelegatorAddr, msg.ValidatorAddr)
 	if !found {
-		return ErrNoValidatorForAddress(k.codespace).Result()
+		return ErrNoUnbondingDelegation(k.Codespace()).Result()
 	}
 
-	if ctx.IsCheckTx() {
-		return sdk.Result{}
+	// ensure that enough time has passed
+	ctxTime := ctx.BlockHeader().Time
+	if ubd.MinTime > ctxTime {
+		return ErrNotMature(k.Codespace(), "unbonding", "unit-time", ubd.MinTime, ctxTime).Result()
 	}
 
-	// retrieve the amount of bonds to remove (TODO remove redundancy already serialized)
-	if msg.Shares == "MAX" {
-		delShares = bond.Shares
-	}
+	k.CoinKeeper().AddCoins(ctx, ubd.DelegatorAddr, sdk.Coins{ubd.Balance})
+	k.RemoveUnbondingDelegation(ctx, ubd)
 
-	// subtract bond tokens from delegator bond
-	bond.Shares = bond.Shares.Sub(delShares)
+	tags := sdk.NewTags(
+		tags.Action, ActionCompleteUnbonding,
+		tags.Delegator, []byte(msg.DelegatorAddr.String()),
+		tags.SrcValidator, []byte(msg.ValidatorAddr.String()),
+	)
 
-	// remove the bond
-	revokeValidator := false
-	if bond.Shares.IsZero() {
+	return sdk.Result{Tags: tags}
+}
 
-		// if the bond is the owner of the validator then
-		// trigger a revoke validator
-		if bytes.Equal(bond.DelegatorAddr, validator.Owner) &&
-			validator.Revoked == false {
-			revokeValidator = true
-		}
+func handleMsgBeginRedelegate(ctx sdk.Context, msg types.MsgBeginRedelegate, k keeper.Keeper) sdk.Result {
 
-		k.removeDelegation(ctx, bond)
-	} else {
-		// Update bond height
-		bond.Height = ctx.BlockHeight()
-		k.setDelegation(ctx, bond)
+	returnAmount, err := k.Unbond(ctx, msg.DelegatorAddr, msg.ValidatorSrcAddr, msg.SharesAmount)
```

Comment:
Do we prevent transitive redelegation (redelegating from A to B, then again from B to C before the first redelegation has completed)? I think transitive redelegation will break some accounting, both here and in slashing.